### PR TITLE
check quota and allow updating it

### PR DESCRIPTION
**Pull request by michaelmattig**
- [ ] I added an entry to [`CHANGELOG.md`](CHANGELOG.md) if knowledge of this change could be valuable to users.

---

Here is a brief summary of what I did:

<TEXT>

### CODE REVIEW

**Comment by michaelmattig** [comment]
[Credit-System: User Credits Einsatz](https://trello.com/c/b95KAN0M/1178-credit-system-user-credits-einsatz)

File: services/src/handlers/plots.rs [hunk 1]
```diff
@@ -179,7 +181,10 @@ async fn get_plot_handler<C: Context>(
         });
     };
 
-    let processor = initialized.query_processor().context(error::Operator)?;
+    let processor = execution_context
+        .create_plot_query_processor(initialized)
+        .await
+        .context(error::Operator)?;
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/plots.rs:187]: 1. What happens if we call it like before?
2. Why is it wrapped here, but all other methods just take a reference to the execution context?

File: services/src/pro/util/tests.rs [hunk 2]
```diff
@@ -296,3 +310,87 @@ pub(in crate::pro) mod mock_oidc {
         result
     }
 }
+
+#[allow(clippy::missing_panics_doc)]
+pub async fn register_ndvi_workflow_helper(ctx: &ProInMemoryContext) -> (Workflow, WorkflowId) {
```
Comment (ChristianBeilschmidt) [inline services/src/pro/util/tests.rs:315]: Just as a thought:
Should we only compile this if(test)?

Maybe clippy would not bother then about the missing panics doc.
And we could not use it apart from test.

File: services/src/pro/users/postgres_userdb.rs [hunk 3]
```diff
@@ -66,17 +66,23 @@ where
 
         let stmt = tx
             .prepare(
-                "INSERT INTO users (id, email, password_hash, real_name, active) VALUES ($1, $2, $3, $4, $5);",
+                "INSERT INTO users (id, email, password_hash, real_name, quota_available, active) VALUES ($1, $2, $3, $4, $5, $6);",
```
Comment (ChristianBeilschmidt) [inline services/src/pro/users/postgres_userdb.rs:69]: If there is a quote no matter what is specified if it is used, shouldn't we set it explicitly to 0 in the settings. Currently, it is commented out.

File: services/src/pro/users/hashmap_userdb.rs [hunk 4]
```diff
@@ -56,6 +58,14 @@ impl UserDb for HashMapUserDb {
 
         self.users.write().await.insert(id.to_string(), user);
 
+        let mut quota_available = self.quota_available.write().await;
+        quota_available.insert(
+            id,
+            crate::util::config::get_config_element::<crate::pro::util::config::User>()?
+                .default_available_quota
+                .unwrap_or(0),
```
Comment (ChristianBeilschmidt) [inline services/src/pro/users/hashmap_userdb.rs:66]: maybe already deserliaze it as 0 if unset

File: services/src/pro/contexts/mod.rs [hunk 5]
```diff
@@ -137,6 +183,30 @@ where
...
+        operator.query_processor()
+    }
+
+    async fn create_plot_query_processor(
+        &self,
+        operator: Box<dyn InitializedPlotOperator>,
+    ) -> geoengine_operators::util::Result<TypedPlotQueryProcessor> {
+        check_quota(&*self.user_db, &self.session).await?;
+        operator.query_processor()
+    }
```
Comment (ChristianBeilschmidt) [inline services/src/pro/contexts/mod.rs:208]: Is this the right place?
Shouldn't it be checked when we call `query`?

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: services/src/handlers/plots.rs (hunk 1 above)
Comment (michaelmattig) [reply services/src/handlers/plots.rs:187]: 1. If you create the processor from the InitializedOperator, like before, then no quota check is performed. This here is basically the point where, in Pro, the quota is checked, and the execution prevented if it is less than zero. In free, no check is performed, because there is no quota.

2. I did it this way to implement a different behavior, depending on the context being Pro or Free. That#s why `create_plot_query_processor` is a trait method with different implementations depending on the type of context.

I am sure there are plenty other ways to implement the quota check. E.g. duplicating the WMS handler code in Pro, and including a check there. But I found my solution to be the easiest, least intrusive and most effective.

File: services/src/pro/util/tests.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/pro/util/tests.rs:315]: Stackoverflow says that the Rust compiler doesn't include unused methods in the binary. So I don't think we need to do this. But we could.

File: services/src/pro/users/postgres_userdb.rs (hunk 3 above)
Comment (michaelmattig) [reply services/src/pro/users/postgres_userdb.rs:69]: Right now, if no quota is set, it would just default to zero. If we make it non-optional and no quota is set, geo engine will crash at runtime instead, when the first workflow is run. If the latter is our preferred behavior, we can make it non-optional. I thought taking 0 made sense.

File: services/src/pro/users/hashmap_userdb.rs (hunk 4 above)
Comment (michaelmattig) [reply services/src/pro/users/hashmap_userdb.rs:66]: would also be possible but leaving it this way is more flexible right? Because we still know whether it was set to 0 explicitly or left blank.

File: services/src/pro/contexts/mod.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs:208]: During initialization of the query processor we already do some work, like computing ResultDescriptors, initializing models and idk what. So why would we want to do that work and only abort later when finally query is called?

File: services/src/pro/util/tests.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/util/tests.rs:315]: No, this is clear to me. It's just that clippy says, okay, this is a regular method, so why do you just panic?
And it is, we could use it, but we use it only in test. Wondering if we could just enclose it in #[cfg(test)] ?

File: services/src/pro/contexts/mod.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/contexts/mod.rs:208]: Wouldn't you want to run a query as long as you have resources and not to be prevented to run the first query although you would have had resources at least for that?

File: services/src/pro/users/hashmap_userdb.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/users/hashmap_userdb.rs:66]: But this leaves the default value to each implementation, right? It is sufficient, but I just wanted to see if it is actually meant to be.

File: services/src/pro/util/tests.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/pro/util/tests.rs:315]: how about we make the whole file #cfg[test]?

File: services/src/pro/contexts/mod.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs:208]: that's how it currently works: If you have at least 1 credit, you can create the query processor and query it as often as you want.

If we want to check every query-call, we have to do things differently. Like introducing a wrapper operator that checks this in the query-method before delegating it to the real workflow.

File: services/src/pro/users/postgres_userdb.rs [hunk 6]
```diff
@@ -436,7 +451,13 @@ where
     async fn increment_quota_used(&self, user: &UserId, quota_used: u64) -> Result<()> {
         let conn = self.conn_pool.get().await?;
         let stmt = conn
-            .prepare("UPDATE users SET quota_used = quota_used + $1 WHERE id = $2;")
+            .prepare(
+                "
+            UPDATE users SET 
+                quota_available = quota_available - $1,
```
Comment (ChristianBeilschmidt) [inline services/src/pro/users/postgres_userdb.rs:457]: add a test where the quota_available gets negative after a query

File: services/src/handlers/wms.rs [hunk 7]
```diff
@@ -316,7 +316,10 @@ async fn wms_map_handler<C: Context>(
             Box::new(irp)
         };
 
-        let processor = initialized.query_processor().context(error::Operator)?;
+        let processor = execution_context
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/wms.rs:319]: save_dataset and maybe others (?) are missing

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: Settings-test.toml [hunk 8]
```diff
@@ -28,3 +28,7 @@ client_id = ""
 client_secret = ""
 redirect_uri = ""
 scopes = [ ]
+
+[user]
+quota_check = true
+default_available_quota = 0
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply Settings-test.toml]: 27586db60cfc6f20d0938060f84902b2f3d96ff1

File: services/src/pro/util/tests.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/pro/util/tests.rs:315]: 27586db60cfc6f20d0938060f84902b2f3d96ff1

File: services/src/handlers/wms.rs (hunk 7 above)
Comment (michaelmattig) [reply services/src/handlers/wms.rs]: is now automatically checked because every operator is wrapped: 10bc0646cd8e6938210e1ff24a11e9124fd983af

File: services/src/pro/users/postgres_userdb.rs (hunk 6 above)
Comment (michaelmattig) [reply services/src/pro/users/postgres_userdb.rs:457]: bf8442899e6d3725431a335c3cf664c008a11173

File: services/src/pro/contexts/mod.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs:209]: ok so I built it into the query-method now. We ultimately need it there, because otherwise one could create a processor with 1 credit and use it forever because the quota is not checked again.

So right now, every processor in the chain will check the quota. We discussed caching it instead. The cache would be in the query context. Right now, this is not possible because the query context can be re-used for subsequent queries.

I would propose to make the QueryContext non-reusable. So it should be consumed by the query-method. Then we can cache the quota there, and, we can also have something like a unique query id stored in the context that we can use for stuff like logging and quota tracking.

In order for the QueryContext to be consumed, we have to change the parameter in the query method from a fat pointer to a box though. And: for operators with more than one child, we would need to clone the QueryContext, in oder to put it into each child's query method.

File: operators/src/error.rs [hunk 9]
```diff
@@ -349,6 +353,13 @@ pub enum Error {
         expected: usize,
         found: usize,
     },
+
+    #[cfg(feature = "pro")]
+    QuotaExhausted,
```
Comment (ChristianBeilschmidt) [inline operators/src/error.rs:358]: I guess we said we did not want to have `cfg = pro` all around the code. Could we do something about this?

File: operators/src/pro/meta/quota.rs [hunk 10]
```diff
@@ -34,3 +36,10 @@ impl QuotaTracking {
         let _ = self.quota_sender.send(self.computation); // ignore the Result because the quota receiver should never close the receiving end of the channel
     }
 }
+
+#[async_trait]
+pub trait QuotaCheck {
+    async fn check_quota(&self) -> Result<bool>;
```
Comment (ChristianBeilschmidt) [inline operators/src/pro/meta/quota.rs:42]: It should be more verbose what it is doing. For instance, `has_quota_available` would be an option.

Maybe you can add some doc-comments as well.

File: services/src/pro/contexts/mod.rs [hunk 11]
```diff
@@ -137,6 +184,30 @@ where
         op
     }
 
+    async fn create_raster_query_processor(
+        &self,
+        operator: Box<dyn InitializedRasterOperator>,
+    ) -> geoengine_operators::util::Result<TypedRasterQueryProcessor> {
+        check_quota(&*self.user_db, &self.session).await?;
```
Comment (ChristianBeilschmidt) [inline services/src/pro/contexts/mod.rs:191]: It is checked here and injected into the StatisticsProcessor? Why is this necessary?

File: services/src/pro/contexts/mod.rs [hunk 12]
```diff
@@ -312,6 +393,28 @@ where
...
+            .user_db
+            .quota_available(&self.session)
+            .await
+            .map_err(
+                |e| geoengine_operators::error::Error::QueryingProcessorFailed {
+                    source: Box::new(e),
+                },
+            )?
+            > 0)
+    }
```
Comment (ChristianBeilschmidt) [inline services/src/pro/contexts/mod.rs:415]: Do we need to check (or start doing this…) performance degredations?
This way, we could decide to work on this now or… not.

File: services/src/pro/handlers/drone_mapping.rs [hunk 13]
```diff
@@ -422,6 +423,11 @@ mod tests {
         let ctx = ProInMemoryContext::test_default();
         let session = create_session_helper(&ctx).await;
 
+        ctx.user_db_ref()
+            .update_quota_available_by_user(&session.user.id, 9999)
+            .await
+            .unwrap();
```
Comment (ChristianBeilschmidt) [inline services/src/pro/handlers/drone_mapping.rs:429]: Should we really put this into this test? It is not really part of what we want to test.

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: services/src/pro/contexts/mod.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/contexts/mod.rs:209]: I guess it is already designed this way, so we could do that.

File: services/src/pro/handlers/drone_mapping.rs (hunk 13 above)
Comment (michaelmattig) [reply services/src/pro/handlers/drone_mapping.rs:429]: by default, in all tests, the available quota is zero. We could change it and have by default 9999 and only set it to zero in the tests about quotas

File: services/src/pro/contexts/mod.rs [hunk 14]
```diff
@@ -54,55 +57,99 @@ pub trait ProContext: Context<Session = UserSession> {
...
+        geoengine_operators::error::Error::CreatingProcessorFailed {
+            source: Box::new(e),
+        }
+    })?;
+
+    if quota_available <= 0 {
+        return Err(geoengine_operators::error::Error::CreatingProcessorFailed {
+            source: Box::new(crate::error::Error::QuotaExhausted),
+        });
+    }
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs:126]: no, if zero or less is available, it is exhausted

File: services/src/pro/contexts/mod.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs:209]: the problem here is, if the QueryContext is cloneable, there is no benefit in it being consumed by the query method. You can just clone it before you put it in. So we would somehow have to invalidate it after being put into a query-method, but put it in again, for all the child processors.

I guess then the best idea would be to avoid wrapping every single processor with the quota check, but only wrap the top-level one. But how can the execution context distinguish the top-level operator from the other ones? Maybe we have to do the top-level wrapping in a different place.

One idea would be to do it in the `Workflow`. Right now, we access the top-level operator by `workflow.operator` directly. We could make `Workflow` into a trait with a method `operator()` which, in free just outputs the operator but in pro it wraps it with a quota checker.

File: services/src/pro/contexts/mod.rs (hunk 11 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs]: it was leftover from the previous attempt. I removed it

File: services/src/pro/contexts/mod.rs (hunk 12 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs]: I performed a benchmark (see my Trello comment)

>     Starting benchmark, quota check enabled: false
    Run 0 time 249.679598ms
    Run 1 time 242.757376ms
    Run 2 time 243.797694ms
    Run 3 time 247.024981ms

>    Starting benchmark, quota check enabled: true
    Run 0 time 255.951031ms
    Run 1 time 251.298686ms
    Run 2 time 251.200896ms
    Run 3 time 245.754991ms

File: services/src/pro/handlers/users.rs [hunk 15]
```diff
@@ -276,18 +282,21 @@ pub(crate) async fn quota_handler<C: ProContext>(
     ctx: web::Data<C>,
     session: C::Session,
 ) -> Result<impl Responder> {
-    let quota_used = ctx.user_db_ref().quota_used(&session).await?;
-    Ok(web::Json(QuotaUsed { used: quota_used }))
+    let available = ctx.user_db_ref().quota_available(&session).await?;
+    let used = ctx.user_db_ref().quota_used(&session).await?;
+
+    Ok(web::Json(Quota { available, used }))
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/pro/handlers/users.rs:288]: for this to work, Quota should have to implement Responder, which it doesn't

>   .route(web::get().to(get_user_quota_handler::<C>))
    |                                   -- ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `actix_web::Responder` is not implemented for `pro::handlers::users::Quota`
    |                                   |
    |                                   required by a bound introduced by this call

File: services/src/pro/handlers/users.rs [hunk 16]
```diff
@@ -309,13 +318,52 @@ pub(crate) async fn user_quota_handler<C: ProContext>(
...
+                source: Box::new(error::Error::OperationRequiresAdminPrivilige),
             });
         }
     }
 
-    let quota_used = ctx.user_db_ref().quota_used_by_user(&user).await?;
-    Ok(web::Json(QuotaUsed { used: quota_used }))
+    let available = ctx.user_db_ref().quota_available_by_user(&user).await?;
+    let used = ctx.user_db_ref().quota_used_by_user(&user).await?;
+    Ok(web::Json(Quota { available, used }))
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/pro/handlers/users.rs:328]: for this to work, Quota should have to implement Responder, which it doesn't

>   .route(web::get().to(get_user_quota_handler::<C>))
    |                                   -- ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `actix_web::Responder` is not implemented for `pro::handlers::users::Quota`
    |                                   |
    |                                   required by a bound introduced by this call

File: services/src/pro/handlers/users.rs [hunk 17]
```diff
@@ -309,13 +318,52 @@ pub(crate) async fn user_quota_handler<C: ProContext>(
...
+) -> Result<impl Responder> {
+    let user = user.into_inner();
+
+    let update = update.into_inner();
+
+    ctx.user_db_ref()
+        .update_quota_available_by_user(&user, update.available)
+        .await?;
+
+    Ok(actix_web::HttpResponse::Ok())
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/pro/handlers/users.rs:366]: for this to work, Quota should have to implement Responder, which it doesn't

>   .route(web::get().to(get_user_quota_handler::<C>))
    |                                   -- ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `actix_web::Responder` is not implemented for `pro::handlers::users::Quota`
    |                                   |
    |                                   required by a bound introduced by this call

File: operators/src/error.rs (hunk 9 above)
Comment (michaelmattig) [reply operators/src/error.rs]: I removed the feature flag. It doesn't hurt to have the enum variant in non-pro and we do it this way all the time, e.g. with permission exceptions in the services crate. 3c7babbdf87ddc47574187df72dcdb57a808070b

File: operators/src/pro/meta/statistics.rs [hunk 18]
```diff
@@ -155,7 +164,7 @@ where
         match stream_result {
             Ok(stream) => {
                 tracing::debug!(event = "query ok");
-                Ok(StreamStatisticsAdapter::new(stream, span.clone(), quota).boxed())
+                Ok(StreamStatisticsAdapter::new(stream, span.clone(), quota_tracker).boxed())
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/pro/meta/statistics.rs:164]: I renamed the operator and processor to a generic name. 3c7babbdf87ddc47574187df72dcdb57a808070b

File: services/src/pro/contexts/mod.rs [hunk 19]
```diff
@@ -54,55 +57,99 @@ pub trait ProContext: Context<Session = UserSession> {
...
             layer_provider_db,
+            user_db,
             thread_pool,
             session,
             tiling_specification,
         }
     }
 }
 
+async fn check_quota<U: UserDb>(
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs]: the method was redundant and I removed it.

File: operators/src/error.rs (hunk 9 above)
Comment (ChristianBeilschmidt) [reply operators/src/error.rs]: But now there is a QuotaError also for non-pro, right?

File: operators/src/pro/meta/quota.rs (hunk 10 above)
Comment (ChristianBeilschmidt) [reply operators/src/pro/meta/quota.rs]: I don't see in in this change, maybe you added the wrong one?

File: services/src/pro/handlers/drone_mapping.rs (hunk 13 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/handlers/drone_mapping.rs]: That sounds about right. I don't know about the 9999 part. Is this the maximum quote we can think of? :smile:

File: services/src/pro/handlers/users.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply services/src/pro/handlers/users.rs:288]: Okay, but shouldn't we implement that then? Using it as some generic responder doesn't make this better, right?

File: services/src/pro/contexts/mod.rs (hunk 12 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/contexts/mod.rs]: So we leave it there for now?

File: operators/src/error.rs (hunk 9 above)
Comment (michaelmattig) [reply operators/src/error.rs]: Please make a suggestion then.

File: operators/src/pro/meta/quota.rs (hunk 10 above)
Comment (michaelmattig) [reply operators/src/pro/meta/quota.rs]: a845e97cc0954aa90fb5c2acd53d1cda62c98a5f

File: services/src/pro/contexts/mod.rs (hunk 12 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs]: I'd say so

File: services/src/pro/handlers/drone_mapping.rs (hunk 13 above)
Comment (michaelmattig) [reply services/src/pro/handlers/drone_mapping.rs]: it's big enough for the tests and short enough to be read

File: services/src/pro/handlers/users.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/pro/handlers/users.rs:288]: > Okay, but shouldn't we implement that then? Using it as some generic responder doesn't make this better, right?

I wouldn't want to implement this trait for every single struct we return. What's the problem with web::Json all of a sudden? We use it like this all the time.

File: operators/src/error.rs (hunk 9 above)
Comment (ChristianBeilschmidt) [reply operators/src/error.rs]: I don't understand this. I always advocate not mixing the pro/non-pro code.

File: services/src/pro/handlers/users.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply services/src/pro/handlers/users.rs:288]: This is why I suggested using `web::Json<Quota>`.

File: services/src/pro/handlers/drone_mapping.rs (hunk 13 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/handlers/drone_mapping.rs]: If it should never fail, maybe int max?

File: operators/src/error.rs (hunk 9 above)
Comment (michaelmattig) [reply operators/src/error.rs]: > I don't understand this. I always advocate not mixing the pro/non-pro code.

I'm all for that

> This is the only #[cfg(feature = "pro")] that was there.

true. That's why I removed it. I just meant that there are other errors, that are only ever used when a feature is activated (e.g. pro) that are not conditionally compiled.

> What about putting the QuotaError inside the QueryingProcessorFailed or something similar?

that's what I meant here: " I have no other solution here, other than using a generic Box error.". So a new variant `QueryingProcessorFailed { source: Box<dyn Error> }` to hide the real error from pro.

Or do you have a better idea?

File: services/src/pro/handlers/drone_mapping.rs (hunk 13 above)
Comment (michaelmattig) [reply services/src/pro/handlers/drone_mapping.rs]: that's a bulky number, but if you insist I can change it. which int would you like? 32 or 64 bit? signed or unsigned?

File: services/src/pro/handlers/users.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/pro/handlers/users.rs:328]: 58cd9244c5445fc15d50c454f0863b11d7d0d7ed

File: services/src/pro/handlers/users.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/pro/handlers/users.rs]: 58cd9244c5445fc15d50c454f0863b11d7d0d7ed

File: operators/src/error.rs (hunk 9 above)
Comment (ChristianBeilschmidt) [reply operators/src/error.rs]: No, I guess we have to do it this way.

We just have to get sure that we still get a reasonable error out.

Comment (michaelmattig) [reply operators/src/error.rs]: 89751bd532aead2d2f0d99f81480fed7038de92d

**Review by ChristianBeilschmidt — APPROVED** [review]

**Comment by ChristianBeilschmidt** [comment]
bors merge
